src/array.cr

```diff
@@ -751,6 +751,11 @@ class Array(T)
   # a.delete_at(99, 1) # raises IndexError
   # ```
   def delete_at(index : Int, count : Int)
+    index += size if index < 0
+    unless 0 <= index <= size
```

Comment:
Shouldn't we disallow `index == size`? 

There other places where `raise IndexError.new unless 0 <= index <= size` is used is because `index == size` is a valid one.

Reply:
Check `Array#[]` and specifically `Array#[]?(start : Int, count : Int)`. If you take elements right where it ends, there are no elements to be returned and this is valid. I copied that behavior from there. `Array#delete_at` works like that too:

https://play.crystal-lang.org/#/r/971l

If we want to do it differently, we should change all of those methods, but I wouldn't do it in this PR (and I think the current behavior is fine).

Reply:
Also check Ruby:

```
irb(main):001:0> "foo"[3..10]
=> ""
irb(main):002:0> "foo"[4..10]
=> nil
irb(main):003:0> a = [1, 2, 3]
irb(main):004:0> a[3..10]
=> []
irb(main):005:0> a[4..10]
=> nil
```

Reply:
I agree with the behavior of `Array#[]=` . I wasn't sure we wanted the for delete_at since we are explicitly removing and we will know there is nothing.

But as far as this was a double check of the intention I'm fine.

Reply:
Got it. Yeah, I'm not entirely sure about that behavior, I think it makes sense. And in Ruby this is called `slice!`. Here we go:

```
irb(main):011:0> a = [1, 2, 3]
irb(main):012:0> a.slice!(3, 10)
=> []
irb(main):013:0> a
=> [1, 2, 3]
irb(main):014:0> a.slice!(4, 10)
=> nil
```

Reply:
(`slice!` returns the slice that was removed, and `nil` is returned when the index is out of bounds)